orchagent/aclorch.cpp

```diff
@@ -157,18 +166,320 @@ static acl_ip_type_lookup_t aclIpTypeLookup =
+{
+    return m_name;
+}
+
+const set<sai_acl_bind_point_type_t>& AclTableType::getBindPointTypes() const
+{
+    return m_bpointTypes;
+}
+
+const map<sai_acl_table_attr_t, shared_ptr<AclTableMatchInterface>>& AclTableType::getMatches() const
+{
+    return m_matches;
+}
+
+const set<sai_acl_action_type_t>& AclTableType::getActions() const
+{
+    return m_aclAcitons;
+}
+
+AclTableTypeBuilder& AclTableTypeBuilder::withName(string name)
+{
+    m_tableType.m_name = name;
+    return *this;
+}
+
+AclTableTypeBuilder& AclTableTypeBuilder::withBindPointType(sai_acl_bind_point_type_t bpointType)
+{
+    m_tableType.m_bpointTypes.insert(bpointType);
+    return *this;
+}
+
+AclTableTypeBuilder& AclTableTypeBuilder::withMatch(shared_ptr<AclTableMatchInterface> match)
+{
+    m_tableType.m_matches.emplace(match->getId(), match);
+    return *this;
+}
+
+AclTableTypeBuilder& AclTableTypeBuilder::withAction(sai_acl_action_type_t action)
+{
+    m_tableType.m_aclAcitons.insert(action);
+    return *this;
+}
+
+AclTableType AclTableTypeBuilder::build()
+{
+    auto tableType = m_tableType;
+    m_tableType = AclTableType();
+    return tableType;
+}
+
+bool AclTableTypeParser::parse(const std::string& key,
+                               const vector<swss::FieldValueTuple>& fieldValues,
+                               AclTableTypeBuilder& builder)
+{
+    builder.withName(key);
+
+    for (const auto& fieldValue: fieldValues)
+    {
+        auto field = to_upper(fvField(fieldValue));
+        auto value = to_upper(fvValue(fieldValue));
+
+        SWSS_LOG_DEBUG("field %s, value %s", field.c_str(), value.c_str());
+
+        if (field == ACL_TABLE_TYPE_MATCHES)
+        {
+            if (!parseAclTableTypeMatches(value, builder))
+            {
+                return false;
+            }
+        }
+        else if (field == ACL_TABLE_TYPE_ACTIONS)
+        {
+            if (!parseAclTableTypeActions(value, builder))
+            {
+                return false;
+            }
+        }
+        else if (field == ACL_TABLE_TYPE_BPOINT_TYPES)
+        {
+            if (!parseAclTableTypeBindPointTypes(value, builder))
+            {
+                return false;
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown field %s: value %s", field.c_str(), value.c_str());
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool AclTableTypeParser::parseAclTableTypeMatches(const std::string& value, AclTableTypeBuilder& builder)
+{
+    auto matches = tokenize(value, comma);
+    set<sai_acl_range_type_t> saiRangeTypes;
+
+    for (const auto& match: matches)
+    {
+        auto matchIt = aclMatchLookup.find(match);
+        auto rangeMatchIt = aclRangeTypeLookup.find(match);
+
+        if (rangeMatchIt != aclRangeTypeLookup.end())
+        {
+            auto rangeType = rangeMatchIt->second;
+            saiRangeTypes.insert(rangeType);
+        }
+        else if (matchIt != aclMatchLookup.end())
```

Comment:
Why is it ```else```? Can we match both ```ports``` and some other fields?

Reply:
This is mutually exclusive. We either found the match string in aclRangeTypeLookup and then we configure table with AclTableRangeMatch (SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE), or we found a regular aclMatchLookup and configure with AclTableMatch.

Reply:
For example, can we match src_ip and dst_port range at an ACL rule?

Reply:
Yes we can and there are tests for it. In this loop, in case ACL_TABLE_TYPE, "matches" variable is ["SRC_IP", "DST_PORT_RANGE"]